ocaml/parsing/lexer.mll

```diff
@@ -128,6 +128,97 @@ let is_in_string = ref false
 let in_string () = !is_in_string
 let print_warnings = ref true
 
+type deferred_token =
+  { token : token
+  ; start_pos : Lexing.position
+  ; end_pos : Lexing.position
+  }
+
+let deferred_tokens : deferred_token Queue.t = Queue.create ()
+
+(* Effectively splits the current lexer production into two halves. The current
+   call to the lexer will return the first half of the text matched by the
+   production, and the next call to the lexer will return the second half (of
+   length [len]) of the text matched by the production.
+
+   It accomplishes this by doing two things:
+    - It sets the current bounds of the lexbuf to only account for the
+      first half of the text. (The first half is of length: |text|-len.)
+    - It sets the [next_token] ref such that, the next time the lexer is
+      called, it will return the specified [token] *and* set the bounds of
+      the lexbuf to account for the second half of the text. (The second half
+      is of length: |text|.)
+
+   This business with setting the bounds of the lexbuf is only so that error
+   messages point at the right place in the program text.
+*)
+let enqueue_token_from_end_of_lexbuf (lexbuf : Lexing.lexbuf) token ~len =
```

Comment:
As discussed in person, I think the terminology "end_of_lexbuf" is a little confusing because there is lots more stuff in the lexbuf.  Possibly `enqueue_token_from_end_of_window` with a definition of window somewhere would be fine.  I also think `enqueue_token_at_start_of_lexbuf` would make sense (as this token is about to be the moral start of the lexbuf).